### PR TITLE
[MC-1361] Add Bitwarden plugin

### PR DESCRIPTION
**Pull request by dsliwinski-r7**
### Description

Describe the proposed changes:

  - Add Bitwarden plugin

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: createMember\n[ACTION] Creating a new member...\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "member": {
        "accessAll": true,
        "email": "user@example.com",
        "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
        "object": "member",
        "resetPasswordEnrolled": false,
        "status": "0-Invited",
        "twoFactorEnabled": false,
        "type": "2-User"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/createMember.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: deleteMember\n[ACTION] Deleting member 44d88612-fea8-a8f3-6de8-2e1278abb02f...\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/deleteMember.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: listAllCollections\n[ACTION] Getting a list of collections...\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "collections": [
        {
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "object": "collection"
        },
        {
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "object": "collection"
        },
        {
          "externalId": "collection_2_external_id",
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "object": "collection"
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/listAllCollections.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: listAllGroups\n[ACTION] Getting a list of groups...\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "groups": [
        {
          "accessAll": false,
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "name": "Test Group 1",
          "object": "group"
        },
        {
          "accessAll": true,
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "name": "Development Team Group",
          "object": "group"
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/listAllGroups.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: listAllMembers\n[ACTION] Getting a list of members...\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "members": [
        {
          "accessAll": true,
          "email": "user@example.com",
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "name": "User",
          "object": "member",
          "resetPasswordEnrolled": false,
          "status": "2-Confirmed",
          "twoFactorEnabled": false,
          "type": "0-Owner",
          "userId": "f4aac844-045f-41ad-9547-af800161c946"
        },
        {
          "accessAll": true,
          "email": "user2@example.com",
          "externalId": "external_id_123456",
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "name": "User 2",
          "object": "member",
          "resetPasswordEnrolled": false,
          "status": "2-Confirmed",
          "twoFactorEnabled": false,
          "type": "0-Owner",
          "userId": "ccfe60d0-136a-41ee-8669-af8100d0de6e"
        },
        {
          "accessAll": true,
          "email": "user3@example.com",
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "object": "member",
          "resetPasswordEnrolled": false,
          "status": "0-Invited",
          "twoFactorEnabled": false,
          "type": "2-User"
        },
        {
          "accessAll": true,
          "email": "user4@example.com",
          "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "object": "member",
          "resetPasswordEnrolled": false,
          "status": "0-Invited",
          "twoFactorEnabled": false,
          "type": "2-User"
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/listAllMembers.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Bitwarden:1.0.0. Step name: listEvents\n[ACTION] Getting a list of events...\n[API] Getting authentication token...\n",
    "status": "ok",
    "meta": {},
    "output": {
      "events": [
        {
          "object": "event",
          "type": 1503,
          "memberId": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "date": "2023-01-31T08:11:49.2009698Z",
          "ipAddress": "198.51.100.0"
        },
        {
          "object": "event",
          "type": 1504,
          "memberId": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "date": "2023-01-31T08:10:30.9437611Z",
          "ipAddress": "198.51.100.0"
        },
        {
          "object": "event",
          "type": 1502,
          "memberId": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "date": "2023-01-31T08:09:08.5691696Z",
          "ipAddress": "198.51.100.0"
        },
        {
          "object": "event",
          "type": 1500,
          "memberId": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
          "date": "2023-01-31T07:58:22.8964036Z",
          "ipAddress": "198.51.100.0"
        }
      ]
    }
  },
  "version": "v1",
  "type": "action_event"
}
```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/listEvents.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: reinviteMember\n[ACTION] Reinviting a member: 44d88612-fea8-a8f3-6de8-2e1278abb02f\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/reinviteMember.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: retrieveMember\n[ACTION] Getting information of member id: 44d88612-fea8-a8f3-6de8-2e1278abb02f\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "member": {
        "accessAll": true,
        "email": "user@example.com",
        "externalId": "external_id_123456",
        "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
        "name": "User",
        "object": "member",
        "resetPasswordEnrolled": false,
        "status": "2-Confirmed",
        "twoFactorEnabled": false,
        "type": "0-Owner",
        "userId": "44d88612-fea8-a8f3-6de8-2e1278abb02f"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/retrieveMember.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: retrieveMembersGroupIds\n[ACTION] Getting group IDs for a member: 44d88612-fea8-a8f3-6de8-2e1278abb02f\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "membersGroupIds": [
        "44d88612-fea8-a8f3-6de8-2e1278abb02f",
        "44d88612-fea8-a8f3-6de8-2e1278abb02f"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/retrieveMembersGroupIds.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: updateMember\n[ACTION] Updating details for member 44d88612-fea8-a8f3-6de8-2e1278abb02f...\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "member": {
        "accessAll": true,
        "email": "user@example.com",
        "externalId": "after-update-id",
        "id": "44d88612-fea8-a8f3-6de8-2e1278abb02f",
        "object": "member",
        "resetPasswordEnrolled": false,
        "status": "0-Invited",
        "twoFactorEnabled": false,
        "type": "2-User"
      }
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/updateMember.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/https://example.com Step name: updateMembersGroups\n[ACTION] Updating associated groups for a member: 44d88612-fea8-a8f3-6de8-2e1278abb02f\n[API] Getting authentication token...\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/bitwarden:1.0.0 --debug run < tests/updateMembersGroups.json
</summary>
</details>

### CODE REVIEW

**Review by cdankowski-r7 — APPROVED** [review]

**Review by mgrandys-r7 — APPROVED** [review]

**Comment by dlaverty-r7** [comment]
@dsliwinski-r7  Please can you take a look at the plugin validation errors (outside of the missing keyword)?

**Comment by dsliwinski-r7** [comment]
> @dsliwinski-r7 Please can you take a look at the plugin validation errors (outside of the missing keyword)?

@dlaverty-r7 `HelpInputOutputValidator` fails because it incorrectly recognizes the given timestamp examples in help.md as invalid. This is a known issue with the PluginValidator.

File: plugins/bitwarden/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+FROM rapid7/insightconnect-python-3-38-plugin:4
```
Comment (llaszuk-r7) [inline plugins/bitwarden/Dockerfile:1]: Can you use a slim version for the docker image?

**Review by llaszuk-r7 — APPROVED** [review]
LGTM just one comment about docker image version

**Review by dlaverty-r7 — APPROVED** [review]
Lgtm - can you address Lukasz's comment re slim version and we can get this one merged @dsliwinski-r7 ?

**Comment by dsliwinski-r7** [comment]
> Lgtm - can you address Lukasz's comment re slim version and we can get this one merged @dsliwinski-r7 ?

@dsliwinski-r7 Sure, I will change the version. To be sure, should this plugin be set as a cloud plugin in plugin spec? :)

**Comment by dlaverty-r7** [comment]
> > Lgtm - can you address Lukasz's comment re slim version and we can get this one merged @dsliwinski-r7 ?
> 
> @dsliwinski-r7 Sure, I will change the version. To be sure, should this plugin be set as a cloud plugin in plugin spec? :)

Hi @dsliwinski-r7 , I think he is referring to using the slim base image.

**Review by dlaverty-r7 — APPROVED** [review]

**Comment by dlaverty-r7** [comment]
@dsliwinski-r7 , would you mind rebasing this one please and we'll get it merged?

**Comment by dsliwinski-r7** [comment]
> @dsliwinski-r7 , would you mind rebasing this one please and we'll get it merged?

Done

**Pull request merged**